pkg/kv/kvserver/replica_range_lease.go

```diff
@@ -1245,7 +1245,6 @@ func (r *Replica) redirectOnOrAcquireLeaseForRequest(
 	// Loop until the lease is held or the replica ascertains the actual lease
 	// holder. Returns also on context.Done() (timeout or cancellation).
 	for attempt := 1; ; attempt++ {
-		now = r.store.Clock().NowAsClockTimestamp()
 		llHandle, status, transfer, pErr := func() (*leaseRequestHandle, kvserverpb.LeaseStatus, bool, *kvpb.Error) {
```

Comment:
so what was the (or at least a possible) sequence of events here?

1. we grab `now` above
2. someone bumps the minLeaseProposedTS. Is this someone someone else calling `redirectOnOrAcquireLeaseForRequest` and racing ahead of us, which (during eval) hits `RevokeLease`[^1]?
3. but then we would join the ongoing request already, no? And that would prevent us from calling `VerifyAndBuild`, so we wouldn't error out?

[^1]: https://github.com/cockroachdb/cockroach/blob/90d1c8caa24033e6a2f77c768dab0378fa31f514/pkg/kv/kvserver/replica_range_lease.go#L1030-L1036

Reply:
> Is this someone someone else calling redirectOnOrAcquireLeaseForRequest and racing ahead of us, which (during eval) hits RevokeLease[1](https://github.com/cockroachdb/cockroach/pull/135042#user-content-fn-1-53155447d858ae7b5fdcf644d14e8687)?

Yes, I added some additional logging to kvnemesis and confirmed that it is.

> but then we would join the ongoing request already, no? And that would prevent us from calling VerifyAndBuild, so we wouldn't error out?

It looks like the ongoing request completes (unsuccessfully) before the next request acquires the replica mutex. `request lease: err=[NotLeaseHolderError] ‹refusing to acquire lease on follower›`. So there's nothing for it to join to.

This timeline is strange, but there's nothing wrong with it besides the fact that we are using a stale clock timestamp for `now`, which this commit fixes.